README improvements mainly for contributors (#50)
Bonitasoft-Community/Build-Bonita: README.md

- as of Bonita 7.10, tags use the `<bonita_version>-<increment>` like `7.10.0-1`. This allows to track improvements or
bug fixes applied to `Build-Bonita` for a given Bonita version

Examples

| Bonita version | Build-Bonita tag |
| -------- | ----- |
| 7.10.1 | 7.10.1-1, 7.10.1-2, .... |
| 7.10.0 | 7.10.0-1, 7.10.0-2, .... |
| 7.9.4 | 7.9.4 |
| 7.7.5 | 7.7.5 |

# Developing on Build-Bonita

The following is for contributors to this repository.

Notice that a lot of actions are manual, so if it's becoming boring for you, fill an issue to discuss about it, then
provide a Pull Request to automate this and simplify our life

## Add support for a new version

Notice that most of the actions described below can be done directly using the GitHub website, for instance
- file edition
- branch and pull request creation

See [GitHub help](https://help.github.com/en/github/managing-files-in-a-repository/editing-files-in-your-repository) for
more details

### Bonita maintenance version

- create a new branch starting from the `master` branch, for instance `maintenance_7.10.2`
- update the `build-script.sh` file and update the `BONITA_VERSION` variable

> - create a Pull Request targeting the `master` branch (TODO link to github doc)

Author: ```suggestion
- [create a Pull Request](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request) targeting the `master`
```

- wait for build to pass, this should work without any other modifications

> - merge the Pull Request after successful build

Author: ```suggestion
- [merge the Pull Request](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/merging-a-pull-request) after successful build
```

### Bonita development version

- create a new branch starting from the `dev` branch, for instance `dev_7.11.0.W10`
- update the `build-script.sh` file and update the `BONITA_VERSION` variable

> - create a Pull Request targeting the `dev` branch

Author: ```suggestion
- [create a Pull Request](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request) targeting the `dev` branch
```

- wait for build to run
- failures often happen because of new components to be added or removed, build options of some components to be updated
  - try to fix, then commit and iterate until build pass
  - see [#32](https://github.com/Bonitasoft-Community/Build-Bonita/pull/32) or
  [#48](https://github.com/Bonitasoft-Community/Build-Bonita/pull/48) for instance

> - merge the Pull Request after successful build

Author: ```suggestion
- [merge the Pull Request](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/merging-a-pull-request) after successful build
```

### Merging master and dev branches

Follow the same lifecycle as Bonita component repositories. Merge are currently done manually by `Build-Bonita`
contributors
- `master` --> `dev`: all the time, especially after adding support for a new maintenance version. Allow to get new
improvements applied to maintenance versions, avoid subsequent merge conflicts, ...
- `dev` --> `master`: on GA release, `master` is going to become the maintenance branch for the new Bonita released
version. It is highly advised to do the merge in a dedicated branch as some issue occurred at that stage in the past
  - first, ensure that `master` has been merged into`dev`
  - create a new `bonita_7.10.0_GA` branch starting from `master` branch
  - merge `dev` into `bonita_7.10.0_GA`

>   - create a Pull Request from `bonita_7.10.0_GA` targeting the `master` branch

Author: ```suggestion
  - [create a Pull Request](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request) from `bonita_7.10.0_GA` targeting the `master` branch
```

>   - merge the PR into `master` when the build passed (eventually after fixing any issues related to the merge)

Author: ```suggestion
  - [merge the Pull Request](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/merging-a-pull-request) into `master` when the build passed (eventually after fixing any issues related to the merge)
```

## New Release

Release are cut when
- a new Bonita version (GA or maintenance) is supported by `Build-Bonita`

> - significant improvements have been made to t

Reviewer: Little typo here

Author: yes indeed 😈 . I am on it

Author: ```suggestion
- significant improvements have been made in the `Build-Bonita` build script for the latest supported Bonita version
```

> - significant improvements have been made to t

Author: ```suggestion
- significant improvements have been made in the `Build-Bonita` build script for the latest supported Bonita version
```
